packages/Main/src/Layer/VpcLayer.js

```diff
@@ -0,0 +1,341 @@
+
+        const minElevationRanges = [];
+        const maxElevationRanges = [];
+
+        // const resolve = this.addInitializationStep();
+        const resolve = res => res;
+
+        this.loadOctrees = [];
+        this.whenReady = this.source.whenReady.then((/** @type {VpcSource} */ sources) => {
+            this.minElevationRange = this.minElevationRange ?? this.source.minElevation;
+            this.maxElevationRange = this.maxElevationRange ?? this.source.maxElevation;
+
+            sources.forEach((source, i) => {
+                const boundsConforming = source.boundsConforming;
+                const bbox = new THREE.Box3().setFromArray(boundsConforming);
+                const root = {
+                    bbox,
+                    children: [],
+                    sId: i,
+                };
+                const promise =
+                    this.source.sources[i].whenReady.then((src) => {
+                        if (this.source.sources[i].isCopcSource) {
+                            minElevationRanges.push(src.header.min[2]);
+                            maxElevationRanges.push(src.header.max[2]);
+
+                            const { cube, rootHierarchyPage } = src.info;
+                            const { pageOffset, pageLength } = rootHierarchyPage;
+
+                            this.spacing.push(src.info.spacing);
+
+                            const root = new CopcNode(0, 0, 0, 0, pageOffset, pageLength, this, -1, i);
+                            root.bbox.min.fromArray(cube, 0);
+                            root.bbox.max.fromArray(cube, 3);
+                            this.roots[i] = root;
+
+                            return root.loadOctree().then(res => resolve(res));
+                        } else {
+                            minElevationRanges.push(src.boundsConforming[2]);
+                            maxElevationRanges.push(src.boundsConforming[5]);
+
+                            const spacing = (Math.abs(src.bounds[3] - src.bounds[0])
+                                + Math.abs(src.bounds[4] - src.bounds[1])) / (2 * src.span);
+                            this.spacing.push(spacing);
+
+                            const root = new EntwinePointTileNode(0, 0, 0, 0, this, -1, i);
+                            root.bbox.min.fromArray(src.boundsConforming, 0);
+                            root.bbox.max.fromArray(src.boundsConforming, 3);
+                            this.roots[i] = root;
+
+                            return root.loadOctree().then(res => resolve(res));
+                        }
+                    });
+                this.loadOctrees.push(promise);
+
+                root.load = () => this.loadOctrees[i].then(res => res.load());
+                this.roots.push(root);
+            });
+            this.ready = true;
+
+            return this.loadOctrees;
+        });
+    }
+
+    // adapted from PointCloudLayer, return are differents
```

Comment:
A lot of shared code with PointCloudLayer, can yoou factorize it to ease maintenance later on please ?